File: lib/models/membership.dart
/// Membership catalog entry returned by `/membership/tiers`.
class MembershipTier {
  const MembershipTier({
    required this.code,
    required this.name,
    required this.price,
    required this.days,
    required this.features,
  });

  final String code;
  final String name;
  final double price;
  final int days;
  final List<String> features;

  bool get isFree => code == 'free';

  factory MembershipTier.fromJson(Map<String, dynamic> json) => MembershipTier(
    code: (json['code'] as String?) ?? 'free',
    name: (json['name'] as String?) ?? '',
    price: (json['price'] as num?)?.toDouble() ?? 0.0,
    days: (json['days'] as num?)?.toInt() ?? 0,
    features: (json['features'] as List<dynamic>? ?? const [])
        .map((e) => e.toString())
        .toList(),
  );
}

/// A recorded purchase from `/membership/me`.
class MembershipOrder {
  const MembershipOrder({
    required this.id,
    required this.tier,
    required this.price,
    required this.status,
    this.expiresAt,
    this.createdAt,
  });

  final int id;
  final String tier;
  final double price;
  final String status;
  final String? expiresAt;
  final String? createdAt;

  factory MembershipOrder.fromJson(Map<String, dynamic> json) =>
      MembershipOrder(
        id: (json['id'] as num).toInt(),
        tier: (json['tier'] as String?) ?? '',
        price: _asDouble(json['price']),
        status: (json['status'] as String?) ?? 'paid',
        expiresAt: json['expires_at'] as String?,
        createdAt: json['created_at'] as String?,
      );

  /// JSON numbers arrive as num; some drivers serialize DECIMAL as string.
  static double _asDouble(Object? value) {
    if (value is num) return value.toDouble();
    if (value is String) return double.tryParse(value) ?? 0.0;
    return 0.0;
  }
}

File: mobile/src/models/membership.ts
import {
  asDouble,
  asJsonObjectList,
  asNullableString,
  asRequiredInt,
  asString,
  asStringList,
  type JsonObject,
} from '@/json';

/** Membership catalog entry returned by `/membership/tiers`. */
export interface MembershipTier {
  code: string;
  name: string;
  price: number;
  days: number;
  features: string[];
}

export function isFreeTier(tier: MembershipTier): boolean {
  return tier.code === 'free';
}

export function membershipTierFromJson(json: JsonObject): MembershipTier {
  return {
    code: asString(json.code, 'free'),
    name: asString(json.name),
    price: asDouble(json.price),
    days: Math.trunc(asDouble(json.days)),
    features: asStringList(json.features),
  };
}

export function parseTiers(value: unknown): MembershipTier[] {
  return asJsonObjectList(value).map(membershipTierFromJson);
}

/** A recorded purchase from `/membership/me`. */
export interface MembershipOrder {
  id: number;
  tier: string;
  price: number;
  status: string;
  expiresAt: string | null;
  createdAt: string | null;
}

export function membershipOrderFromJson(json: JsonObject): MembershipOrder {
  return {
    id: asRequiredInt(json.id, 'order.id'),
    tier: asString(json.tier),
    price: asDouble(json.price),
    status: asString(json.status, 'paid'),
    expiresAt: asNullableString(json.expires_at),
    createdAt: asNullableString(json.created_at),
  };
}

export function parseOrders(value: unknown): MembershipOrder[] {
  return asJsonObjectList(value).map(membershipOrderFromJson);
}

const TIER_LABELS: Record<string, string> = {
  pro: 'Pro',
  studio: 'Studio',
};

/** Display name for a tier code; unknown codes fall back to `Free`. */
export function tierDisplayName(code: string): string {
  return TIER_LABELS[code] ?? 'Free';
}

const TIER_MEMBER_LABELS: Record<string, string> = {
  pro: 'Pro member',
  studio: 'Studio member',
};

/** Profile-page chip label for a tier code. */
export function tierMemberLabel(code: string): string {
  return TIER_MEMBER_LABELS[code] ?? 'Free';
}
